File: src/android/org/sunbird/sync/SyncPlugin.java
package org.sunbird.sync;

import android.text.TextUtils;

import org.apache.cordova.CallbackContext;
import org.apache.cordova.CordovaInterface;
import org.apache.cordova.CordovaPlugin;
import org.apache.cordova.CordovaWebView;
import org.apache.cordova.PluginResult;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;
import org.sunbird.sync.db.DbService;
import org.sunbird.sync.db.DbServiceImpl;
import org.sunbird.sync.model.HttpResponse;
import org.sunbird.sync.model.NetworkQueueModel;
import org.sunbird.sync.model.Request;
import org.sunbird.sync.network.ApiService;
import org.sunbird.sync.network.ApiServiceImpl;
import org.sunbird.sync.preference.PreferenceService;
import org.sunbird.sync.preference.PreferenceServiceImpl;
import org.sunbird.sync.queue.NetworkQueue;
import org.sunbird.sync.queue.NetworkQueueImpl;

import java.util.ArrayList;

/**
 * This class echoes a string called from JavaScript.
 */
public class SyncPlugin extends CordovaPlugin {

    private static final String TAG = "Cordova-Plugin-SYNC";
    private DbService mDbService;
    private NetworkQueue mNetworkQueue;
    private ApiService mApiService;
    private PreferenceService mPreferenceService;
    private boolean isSyncing;
    private ArrayList<CallbackContext> mHandler = new ArrayList<>();
    private ArrayList<CallbackContext> mLiveHandler = new ArrayList<>();
    private JSONObject mLastEvent;
    private boolean isUnauthorizedErrorThrown;

    @Override
    public void initialize(CordovaInterface cordova, CordovaWebView webView) {
        super.initialize(cordova, webView);
        mDbService = new DbServiceImpl();
        mNetworkQueue = new NetworkQueueImpl(mDbService);
        mApiService = new ApiServiceImpl();
        mPreferenceService = new PreferenceServiceImpl(cordova.getActivity());
    }

    @Override
    public boolean execute(String action, JSONArray args, CallbackContext callbackContext) throws JSONException {
        if (action.equals("sync")) {
            this.syncNetworkQueue(callbackContext);
            return true;
        } else if (action.equals("enqueue")) {
            this.enqueue(args, callbackContext);
            return true;
        } else if (action.equals("onSyncSucces")) {
            mHandler.add(callbackContext);
            return true;
        } else if (action.equals("onAuthorizationError")) {
            mLiveHandler.add(callbackContext);
            return true;
        }

        return false;
    }

    private void syncNetworkQueue(CallbackContext callbackContext) {
        cordova.getThreadPool().execute(new Runnable() {
            public void run() {
                try {
                    mNetworkQueue.seed();
                    while (!mNetworkQueue.isEmpty()) {
                        isSyncing = true;
                        NetworkQueueModel networkQueueModel = mNetworkQueue.peek();
                        HttpResponse httpResponse = mApiService.process(networkQueueModel.getRequest());
                        if (httpResponse != null) {
                            if (httpResponse.getStatus() >= 200 && httpResponse.getStatus() < 300) {
                                handlePostAPIActions(networkQueueModel.getType(), httpResponse);
                                mNetworkQueue.dequeue(false);
                                publishSuccessResult(networkQueueModel, httpResponse);
                            } else if (httpResponse.getStatus() == 400) {
                                publishEvent("error", "BAD_REQUEST");
                                mNetworkQueue.dequeue(true);
                                continue;
                            } else if (httpResponse.getStatus() == 401 || httpResponse.getStatus() == 403) {
                                if (networkQueueModel.getRequest().getNoOfFailureSync() >= 2) {
                                    if(!isUnauthorizedErrorThrown){
                                        isUnauthorizedErrorThrown = true;
                                        if(!isSameToken(networkQueueModel.getRequest(), httpResponse)){
                                            publishAuthErrorEvents(httpResponse.getError(), httpResponse.getStatus());
                                        }
                                        handleUnAuthorizedError(networkQueueModel, httpResponse);
                                        mNetworkQueue.dequeue(true);
                                    } else{
                                        handleUnAuthorizedError(networkQueueModel, httpResponse);
                                        mNetworkQueue.dequeue(true);
                                    }
                                } else {
                                    Request request = networkQueueModel.getRequest();
                                    int noOfFailureSyncs = request.getNoOfFailureSync() + 1;
                                    request.setNoOfFailureSync(noOfFailureSyncs++);
                                    JSONObject model = new JSONObject();
                                    model.put("request", request.toJSON().toString());
                                    mDbService.update("msg_id", new String[]{networkQueueModel.getId()}, model);
                                    handleUnAuthorizedError(networkQueueModel, httpResponse);
                                    mNetworkQueue.dequeue(true);
                                }
                                continue;
                            } else if (httpResponse.getStatus() == -3) {
                                publishEvent(networkQueueModel.getType() + "_error", "NETWORK_ERROR");
                                mNetworkQueue.dequeue(true);
                                break;
                            } else {
                                publishEvent(networkQueueModel.getType() + "_error", httpResponse.getError());
                                mNetworkQueue.dequeue(true);
                                continue;
                            }
                        }
                    }
                    isSyncing = false;
                } catch (Exception e) {
                    e.printStackTrace();
                    isSyncing = false;
                }
            }
        });
    }

    private void handlePostAPIActions(String type, HttpResponse httpResponse) throws JSONException {
        if (type.equalsIgnoreCase("telemetry")) {
            postProcessTelemetrySync(httpResponse);
        }
    }

    public void publishSuccessResult(NetworkQueueModel networkQueueModel, HttpResponse response) throws JSONException {
        JSONObject config = networkQueueModel.getConfig();
        String type = networkQueueModel.getType();
        if (config != null && config.optBoolean("shouldPublishResult")) {
            if (type.equalsIgnoreCase("telemetry")) {
                publishEvent("syncedEventCount", networkQueueModel.getEventCount());
            } else if (type.equalsIgnoreCase("course_progress")) {
                String responseStr = response.getBody();
                JSONObject result = getResultFromAPIResponse(responseStr);
                publishEvent("courseProgressResponse", result);
            } else if (type.equalsIgnoreCase("course_assesment")) {
                String responseStr = response.getBody();
                JSONObject result = getResultFromAPIResponse(responseStr);
                publishEvent("courseAssesmentResponse", result);
            }

        }
    }

    private JSONObject getResultFromAPIResponse(String body) {
        JSONObject result;
        try {
            JSONObject jsonObject = new JSONObject(body);
            result = jsonObject.optJSONObject("result");
        } catch (Exception e) {
            return null;
        }
        return result;
    }

    private void publishEvent(String key, Object value) throws JSONException {
        mLastEvent = new JSONObject();
        mLastEvent.put(key, value);
        consumeEvents();
    }

    private void postProcessTelemetrySync(HttpResponse httpResponse) throws JSONException {
        JSONArray jsonArray = mDbService.read("no_sql", new String[]{"value"}, "key = ?", "last_synced_device_register_is_successful");
        if (jsonArray != null && jsonArray.optJSONObject(0) != null) {
            JSONObject isDeviceRegisterSuccesfullDBObj = jsonArray.optJSONObject(0);
            String isDeviceRegisterSuccesfull = isDeviceRegisterSuccesfullDBObj.optString("value");
            if (isDeviceRegisterSuccesfull.equalsIgnoreCase("false")) {
                if (httpResponse != null) {
                    String responseStr = httpResponse.getBody();
                    try {
                        JSONObject response = new JSONObject(responseStr);
                        if (response != null) {
                            long serverTime = Long.valueOf(response.optString("ets"));
                            long now = System.currentTimeMillis();
                            long currentOffset = serverTime - now;
                            long allowedOffset = Math.abs(currentOffset) > 86400000 ? currentOffset : 0;
                            if (allowedOffset > 0) {
                                JSONObject insertObj = new JSONObject();
                                insertObj.put("key", "telemetry_log_min_allowed_offset_key");
                                insertObj.put("value", String.valueOf(allowedOffset));
                                mDbService.insert(insertObj);
                            }
                        }
                    } catch (Exception e) {
                        e.printStackTrace();
                    }
                }
            }
        }
    }

    private void handleUnAuthorizedError(NetworkQueueModel networkQueueModel, HttpResponse httpResponse) throws JSONException {
        Request request = networkQueueModel.getRequest();
        JSONObject headers = request.getHeaders();
        
        if (isApiTokenExpired(httpResponse.getError(), httpResponse.getStatus())) {
            headers.put("Authorization", "Bearer " + mPreferenceService.getBearerToken());
        } else {
            if (mPreferenceService.getUserToken() != null) {
                headers.put("X-Authenticated-User-Token", mPreferenceService.getUserToken());
            }
            if (mPreferenceService.getManagedUserToken() != null) {
                headers.put("X-Authenticated-For", mPreferenceService.getManagedUserToken());
            }
        }

        request.setHeaders(headers);
        JSONObject model = new JSONObject();
        model.put("request", request.toJSON().toString());
        mDbService.update("msg_id", new String[]{networkQueueModel.getId()}, model);
    }

    private boolean isApiTokenExpired(String error, int statusCode){
        try {
            JSONObject errorObject = new JSONObject(error);
            if (!TextUtils.isEmpty(error)) {
                errorObject = new JSONObject(error);
            }
            return (errorObject != null && "Unauthorized".equalsIgnoreCase(errorObject.optString("message")))
                    || statusCode == 403;
        }catch (JSONException e){
            return false;
        }

    }

    private void enqueue(JSONArray args, CallbackContext callbackContext) {
        cordova.getThreadPool().execute(new Runnable() {
            public void run() {
                try {
                    Object data = args.get(0);
                    JSONObject request = (JSONObject) args.get(1);
                    boolean shouldSync = args.getBoolean(2);
                    String networkRequest = request.getString("request");
                    JSONObject jsonNetworkObject = new JSONObject(networkRequest);
                    jsonNetworkObject.put("body", data);
                    request.put("request", jsonNetworkObject.toString());
                    mDbService.insert(request);
                    if (!isSyncing && shouldSync) {
                        syncNetworkQueue(callbackContext);
                    }
                    callbackContext.success();
                } catch (Exception e) {
                    callbackContext.error(e.getMessage());
                }
            }
        });
    }

    private void consumeEvents() {
        if (this.mHandler.size() == 0 || mLastEvent == null) {
            return;
        }

        for (CallbackContext callback : this.mHandler) {
            final PluginResult result = new PluginResult(PluginResult.Status.OK, mLastEvent);
            result.setKeepCallback(true);
            callback.sendPluginResult(result);
            callback.success(mLastEvent);
        }

        if(mHandler!= null){
            mHandler.clear();
        }
        mLastEvent = null;
    }

    private boolean isSameToken(Request request, HttpResponse httpResponse) {
        String tokenInApp = null;
        String tokenInRequest = null;
        if (isApiTokenExpired(httpResponse.getError(), httpResponse.getStatus())) {
            tokenInApp = "Bearer " + mPreferenceService.getBearerToken();
            tokenInRequest = request.getHeaders().optString("Authorization");
        } else {
            if (mPreferenceService.getUserToken() != null) {
                tokenInApp = mPreferenceService.getUserToken();
                tokenInRequest = request.getHeaders().optString("X-Authenticated-User-Token");
            }
        }
        return (tokenInApp != null && tokenInRequest != null && tokenInApp.equalsIgnoreCase(tokenInRequest));

    }

    private void publishAuthErrorEvents(String error, int statusCode) throws JSONException{
        if (this.mLiveHandler.size() == 0) {
            return;
        }

        JSONObject liveJsonObject = new JSONObject();
        liveJsonObject.put("network_queue_error", isApiTokenExpired(error, statusCode) ? "API_TOKEN_EXPIRED" : "USER_TOKEN_EXPIRED");

        for (CallbackContext callback : this.mLiveHandler) {
            final PluginResult result = new PluginResult(PluginResult.Status.OK, liveJsonObject);
            result.setKeepCallback(true);
            callback.sendPluginResult(result);
        }
    }

}

File: src/ios/SyncPlugin.swift
    import Foundation
    
    @objc(SyncPlugin) class SyncPlugin : CDVPlugin {
        private static var TAG: String = "Cordova-Plugin-SYNC"
        private var mDbService: DbService?
        private var mNetworkQueue: NetworkQueue?
        private var mPreferenceService: PreferenceService?
        private var isSyncing: Bool = false
        private var isUnauthorizedErrorThrown: Bool = false
        private var mTraceId = ""
        private var mLastEvent: [String: Any]? = [:]
        private var mHandler : [String]?
        private var mLiveHandler: [String]?
        
        private func stringifyJSON(_ input: [String: Any]) -> String? {
            do {
                let data = try JSONSerialization.data(withJSONObject: input, options: .prettyPrinted)
                guard let result = String(data: data, encoding: .utf8) else {
                    print("ERROR: failed to cast data as string")
                    return nil
                }
                return result
            } catch let err {
                print("error with " + #function + ": " + err.localizedDescription)
                return nil
            }
        }
        
        override func pluginInitialize() {
            self.mHandler = []
            self.mLiveHandler = []
            self.mDbService = DbServiceImpl()
            self.mNetworkQueue = NetworkQueueImpl(mDbService!)
            self.mPreferenceService = PreferenceServiceImpl()
            if let traceId = mPreferenceService!.getTraceId() {
                self.mTraceId = traceId
            }
        }
        
        @objc
        func sync(_ command: CDVInvokedUrlCommand) {
            let dispatchGroup = DispatchGroup()

            var pluginResult: CDVPluginResult = CDVPluginResult.init(status: CDVCommandStatus_ERROR)
            self.commandDelegate.run(inBackground: {
                self.mNetworkQueue?.seed()
                while !self.mNetworkQueue!.isEmpty() {
                    self.isSyncing = false
                    if let networkQueueModel = self.mNetworkQueue!.peek() {
                        let request = networkQueueModel.getRequest()
                        if let host = request["host"] as? String, let path = request["path"] as? String, let type = request["type"] as? String {
                            let url = URL(string: host + path)!
                            var urlRequest = URLRequest(url: url)
                            urlRequest.httpMethod = type
                            
                            if let requestBody = request["body"]{
                                if let serializer = request["serializer"] as? String {
                                    if serializer.lowercased().elementsEqual("raw") {
                                        urlRequest.httpBody = Data(base64Encoded: requestBody as! String)
                                    } else if serializer.lowercased().elementsEqual("json") {
                                        urlRequest.httpBody = self.stringifyJSON(requestBody as! [String: Any])?.data(using: .utf8)
                                    }
                                }
                            }
                            
                            if let headers = request["headers"] as? [String: String] {
                                for (key, value) in headers {
                                    print(key, value)
                                    // if !["Accept", "Content-Type"].contains(key) {
                                        urlRequest.addValue(value, forHTTPHeaderField: key)
                                    // }
                                }
                            }
                        
                            let task = URLSession.shared.dataTask(with: urlRequest, completionHandler: {(data, response, error) in
                                guard let data = data, let response = response as? HTTPURLResponse else {
                                    print(String(describing: error))
                                    return
                                }
                                do {
                                    let responseJSON = try JSONSerialization.jsonObject(with: data, options: .allowFragments) as? [String: Any]
                                    let statusCode = response.statusCode
                                    if let responseJSON = responseJSON {
                                        if statusCode >= 200 && statusCode <= 300 {
                                            self.handlePostAPIActions(networkQueueModel.getType(), response, responseJSON)
                                            self.mNetworkQueue?.dequeue(false)
                                            self.publishSuccessResult(networkQueueModel, responseJSON)
                                            dispatchGroup.leave()
                                        } else if statusCode == 400 {
                                            self.publishEvent("error", "BAD_REQUEST")
                                            self.mNetworkQueue?.dequeue(true)
                                            dispatchGroup.leave()
                                        } else if statusCode == 403 || statusCode == 401 {
                                            let failedCount = networkQueueModel.getFailedCount()
                                            if failedCount >= 2  {
                                                if !self.isUnauthorizedErrorThrown {
                                                    self.isUnauthorizedErrorThrown = true
                                                    if !self.isSametoken(networkQueueModel, responseJSON, response) {
                                                        self.publishAuthErrorEvents(responseJSON, response.statusCode)
                                                    }
                                                    self.handleUnAuthorizedError(networkQueueModel, response, responseJSON)
                                                    self.mNetworkQueue?.dequeue(true)
                                                } else {
                                                    self.handleUnAuthorizedError(networkQueueModel, response, responseJSON)
                                                    self.mNetworkQueue?.dequeue(true)
                                                }
                                            } else {
                                                let failedCount = networkQueueModel.getFailedCount()
                                                let updatedFailedCount = failedCount + 1
                                                networkQueueModel.setFailedCount(updatedFailedCount)
                                                self.handleUnAuthorizedError(networkQueueModel, response, responseJSON)
                                                var request = networkQueueModel.getRequest()
                                                request["noOfFailureSync"] = updatedFailedCount
                                                self.mDbService?.update("msg_id", [networkQueueModel.getId()], ["request": self.stringifyJSON(request)])
                                                self.mNetworkQueue?.dequeue(true)
                                            }
                                            dispatchGroup.leave()
                                        } else if statusCode == -3 {
                                            let type = networkQueueModel.getType()
                                            self.publishEvent(type + "_error", "NETWORK_ERROR")
                                            self.mNetworkQueue?.dequeue(true)
                                            dispatchGroup.leave()
                                        } else {
                                            let type = networkQueueModel.getType()
                                            //TODO get error from response
                                            self.publishEvent(type + "_error", "NETWORK_ERROR")
                                            self.mNetworkQueue?.dequeue(true)
                                            dispatchGroup.leave()
                                        }
                                    }
                                } catch let error {
                                    print(error)
                                    self.isSyncing = false
                                }
                            })
                            dispatchGroup.enter()
                            task.resume()
                            dispatchGroup.wait()
                        }
                    }
                }
            })
        }
        
        @objc
        func enqueue(_ command: CDVInvokedUrlCommand) {
            var pluginResult: CDVPluginResult = CDVPluginResult.init(status: CDVCommandStatus_ERROR)
            let data = command.arguments[0]
            var request = command.arguments[1] as! [String: Any]
            let shouldSync = command.arguments[2] as! Bool
            guard data != nil && request != nil && shouldSync != nil else {
                self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
                return
            }
            self.commandDelegate.run(inBackground: {
                if let networkRequest = request["request"] as? String {
                    do {
                        var networkRequestObject = try JSONSerialization.jsonObject(with: networkRequest.data(using: .utf8)!, options: .allowFragments) as? [String: Any]
                        
                        if let serializer = networkRequestObject!["serializer"] as? String {
                            if serializer.lowercased().elementsEqual("raw") {
                                networkRequestObject?["body"] = (data as! NSData).base64EncodedString(options: [])
                            } else if serializer.lowercased().elementsEqual("json") {
                                networkRequestObject?["body"] = data as! [String: Any]
                            }
                        }
                        
                        if let networkRequestObjectStringified = self.stringifyJSON(networkRequestObject!) {
                            request["request"] = networkRequestObjectStringified
                            if let _ = self.mDbService?.insert(request) {
                                if shouldSync && !self.isSyncing {
                                    self.sync(command)
                                }
                            }
                        }
                        pluginResult = CDVPluginResult.init(status: CDVCommandStatus_OK)
                        self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
                    } catch let error {
                        print(error)
                        self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
                    }
                }
            })
        }
        
        @objc
        func onSyncSucces(_ command: CDVInvokedUrlCommand) {
            if let _ = self.mHandler {
                self.mHandler?.append(command.callbackId)
            }
        }
        
        @objc
        func onAuthorizationError(_ command: CDVInvokedUrlCommand) {
            if let _ = self.mLiveHandler {
                self.mLiveHandler?.append(command.callbackId)
            }
        }
        
        
        private func publishEvent(_ key: String, _ value: Any) {
            self.mLastEvent = [key: value]
            self.consumeEvents()
        }
        
        private func postProcessTelemetrySync(_ httpResponse: [String: Any]) {
            if let result = self.mDbService?.read("SELECT value from no_sql where key = 'last_synced_device_register_is_successful'") {
                if let row = result.first {
                    let isDeviceRegisterSuccesfullDBObj = row
                    if let isDeviceRegisterSuccesfull = isDeviceRegisterSuccesfullDBObj["value"] as? String {
                        if isDeviceRegisterSuccesfull.lowercased().elementsEqual("false") {
                            if let ets = httpResponse["ets"] {
                                let serverTime = (ets as! NSNumber).uint64Value
                                let now = UInt64(Date().timeIntervalSince1970)
                                let currentOffset = serverTime - now
                                let allowedOffset = currentOffset > 86400000 ? currentOffset : 0
                                if allowedOffset > 0 {
                                    var insertObj: [String: String] = [:]
                                    insertObj["key"] = "telemetry_log_min_allowed_offset_key"
                                    insertObj["value"] = String(allowedOffset)
                                    self.mDbService?.insert(insertObj)
                                }
                            }
                        }
                    }
                }
            }
        }
        
        public func publishSuccessResult(_ networkQueueModel: NetworkQueueModel, _ response: [String: Any]) {
            let config = networkQueueModel.getConfig()
            let type = networkQueueModel.getType()
            if let canPublishResult = config?["shouldPublishResult"] as? Bool {
                if canPublishResult {
                    if type.lowercased().elementsEqual("telemetry") {
                        self.publishEvent("syncedEventCount", networkQueueModel.getEventCount())
                    } else if type.lowercased().elementsEqual("course_progress") {
                        if let result = response["result"] {
                            self.publishEvent("courseProgressResponse", result as Any)
                        }
                    } else if type.lowercased().elementsEqual("course_assesment") {
                        if let result = response["result"] {
                            self.publishEvent("courseAssesmentResponse", result as Any)
                        }
                    }
                }
            }
        }
        
        private func handleUnAuthorizedError(_ networkQueueModel: NetworkQueueModel, _ httpResponse: HTTPURLResponse, _ responseObject: [String: Any]) {
            
            var request = networkQueueModel.getRequest()
            if var requestHeaders = request["headers"] as? [String: Any] {
                if self.isApiTokenExpired(responseObject, httpResponse.statusCode) {
                    if let token = self.mPreferenceService?.getBearerToken() {
                        requestHeaders["Authorization"] = "Bearer " + token
                    }
                }else {
                    if let userToken = self.mPreferenceService?.getUserToken() {
                        requestHeaders["X-Authenticated-User-Token"] = userToken
                    }
                    if let managedUserToken = self.mPreferenceService?.getManagedUserToken() {
                        requestHeaders["X-Authenticated-For"] = managedUserToken
                    }
                }
                
                request["headers"] = requestHeaders
            }
            
            self.mDbService?.update("msg_id", [networkQueueModel.getId()], ["request": request])
            
        }
        
        private func handlePostAPIActions(_ type: String, _ httpResponse: HTTPURLResponse, _ responseObject: [String: Any]) {
            self.setupTraceId(httpResponse)
            if type.lowercased().elementsEqual("telemetry") {
                self.postProcessTelemetrySync(responseObject)
            }
        }
        
        
        private func setupTraceId(_ httpResponse: HTTPURLResponse) {
            var responseTraceId: String?
            if #available(iOS 13.0, *) {
                if let xTraceEnabled = httpResponse.value(forHTTPHeaderField: "X-Trace-Enabled") {
                    responseTraceId = xTraceEnabled
                }
            } else {
                if let headers = httpResponse.allHeaderFields as? [String : Any] {
                    if let xTraceEnabled = headers["X-Trace-Enabled"] as? String {
                        responseTraceId = xTraceEnabled
                    }
                }
            }
            if responseTraceId != nil && self.mTraceId != responseTraceId {
                self.mPreferenceService!.setTraceId(responseTraceId!)
            }
        }
        
        private func consumeEvents() {
            if let mHandler = self.mHandler {
                if mHandler.isEmpty || self.mLastEvent == nil {
                    return
                }
                for callback in mHandler {
                    let pluginResult: CDVPluginResult = CDVPluginResult.init(status: CDVCommandStatus_OK, messageAs: self.mLastEvent)
                    self.commandDelegate.send(pluginResult, callbackId: callback)
                }
                self.mHandler = []
                self.mLastEvent = nil
            }
        }
        
        private func publishAuthErrorEvents(_ responseObject : [String: Any], _ statusCode: Int) {
            if let mLiveHandler = self.mLiveHandler {
                if mLiveHandler.count == 0 {
                    return
                }
                let isApiTokenExpired = self.isApiTokenExpired(responseObject, statusCode)
                let result = ["network_queue_error": isApiTokenExpired ? "API_TOKEN_EXPIRED" : "USER_TOKEN_EXPIRED"]
                for callback in mLiveHandler {
                    let pluginResult: CDVPluginResult = CDVPluginResult.init(status: CDVCommandStatus_OK, messageAs: result)
                    self.commandDelegate.send(pluginResult, callbackId: callback)
                }
            }
        }
        
        private func isSametoken(_ request: NetworkQueueModel, _ responseObject: [String: Any], _ response: HTTPURLResponse) -> Bool {
            var tokenInApp: String?, tokenInRequest: String?
            let request = request.getRequest()
            if let requestHeaders = request["headers"] as? [String: Any] {
                if self.isApiTokenExpired(responseObject, response.statusCode) {
                    if let token = mPreferenceService!.getBearerToken() {
                        tokenInApp = "Bearer " + token
                        tokenInRequest = requestHeaders["Authorization"] as? String
                    }
                } else {
                    if let token = mPreferenceService!.getUserToken(){
                        tokenInApp = token
                        tokenInRequest = requestHeaders["X-Authenticated-User-Token"] as? String
                    }
                }
            }
            
            if tokenInApp != nil && tokenInRequest != nil {
                return tokenInApp!.lowercased().elementsEqual(tokenInRequest!.lowercased())
            }
            
            return false
        }
        
        private func isApiTokenExpired(_ responseObject: [String: Any], _ statusCode: Int) -> Bool {
            var isTokenExpired = false
            if statusCode == 403 {
                isTokenExpired = true
            }
            if let message = responseObject["message"] as? String {
                isTokenExpired = message.lowercased().elementsEqual("unauthorized")
            }
            return isTokenExpired
        }
    }
    
